core/matrix/dense.cpp

```diff
@@ -86,6 +99,29 @@ inline void conversion_helper(Csr<ValueType, IndexType> *result,
     exec->run(op(tmp.get(), source));
     tmp->move_to(result);
 }
+
+
+template <typename ValueType, typename IndexType, typename MatrixType,
+          typename OperationType>
+inline void conversion_helper(Ell<ValueType, IndexType> *result,
+                              MatrixType *source, const OperationType &op)
+{
+    auto exec = source->get_executor();
+
+    Array<size_type> max_nonzeros_per_row(exec, 1);
+    exec->run(TemplatedOperation<ValueType>::
+                  make_calculate_max_nonzeros_per_row_operation(
+                      source, max_nonzeros_per_row.get_data()));
+    size_type padding, max_nnz_per_row;
+    max_nnz_per_row = std::max(result->get_max_nonzeros_per_row(),
+                               *max_nonzeros_per_row.get_data());
```

Comment:
This won't work if `executor` is a GpuExecutor. In this case the pointer returned by `Array::get_data()` is a pointer to GPU memory, so you cannot dereference it on the host. There are two ways you can solve this:

1.  just send a plain reference to a local variable to `calculate_max_nonzeros_per_row`, and leave handling of different memory spaces to the kernel (I would go for this one as it has the benefit of avoiding dynamic memory allocation in CPU and Reference cases)
2.  make sure the array is on the CPU before dereferencing the array (you can move it to the CPU using `max_nonzeros_per_row.set_executor(exec->get_master());`)

Reply:
I realize the same bug was present in the code that was already there, could you please fix that one to? Thanks!

Reply:
Fix them. (the first way)